x-pack/auditbeat/processors/add_session_metadata/add_session_metadata.go

```diff
@@ -0,0 +1,194 @@
+	logName       = "processor." + processorName
+)
+
+var reg *monitoring.Registry
+
+func init() {
+	processors.RegisterPlugin(processorName, New)
+	reg = monitoring.Default.NewRegistry(logName, monitoring.DoNotReport)
+}
+
+type addSessionMetadata struct {
+	config   Config
+	logger   *logp.Logger
+	db       processdb.DB
+	provider provider.Provider
+}
+
+func New(cfg *config.C) (beat.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, fmt.Errorf("fail to unpack the %v configuration: %w", processorName, err)
+	}
+
+	logger := logp.NewLogger(logName)
+
+	ctx := context.TODO()
+	reader := procfs.NewProcfsReader(*logger)
+	db := processdb.NewSimpleDB(reader, *logger)
+
+	backfilledPIDs := db.ScrapeProcfs()
+	logger.Debugf("backfilled %d processes", len(backfilledPIDs))
+
+	switch c.Backend {
+	case "ebpf":
+		p, err := ebpf_provider.NewProvider(ctx, *logger, db)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create ebpf provider: %w", err)
+		}
+		return &addSessionMetadata{
+			config:   c,
+			logger:   logger,
+			db:       db,
+			provider: p,
+		}, nil
+	default:
+		return nil, fmt.Errorf("unknown backend configuration")
+	}
+}
+
+func (p *addSessionMetadata) Run(ev *beat.Event) (*beat.Event, error) {
+	 _, err := ev.GetValue(p.config.PidField)
+	 if err != nil {
+		// Do not attempt to enrich events without PID; it's not a supported event
+		return ev, nil
+	}
+
+	err = p.provider.UpdateDB(ev)
+	if err != nil {
+		return ev, err
+	}
+
+	result, err := p.enrich(ev)
+	if err != nil {
+		return ev, fmt.Errorf("enriching event: %w", err)
+	}
+	return result, nil
+}
+
+func (p *addSessionMetadata) String() string {
+	return fmt.Sprintf("%v=[backend=%s, pid_field=%s, override_fields=%t]",
+		processorName, p.config.Backend, p.config.PidField, p.config.ReplaceFields)
+}
+
+func (p *addSessionMetadata) enrich(ev *beat.Event) (*beat.Event, error) {
+	pidIf, err := ev.GetValue(p.config.PidField)
+	if err != nil {
+		return nil, err
+	}
+	pid, err := pidToUInt32(pidIf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse pid field '%s': %w", p.config.PidField, err)
+	}
+
+	fullProcess, err := p.db.GetProcess(pid)
+	if err != nil {
+		return nil, fmt.Errorf("pid %v not found in db: %w", pid, err)
+	}
+
+	result := ev.Clone()
+
+	processMap := fullProcess.ToMap()
+
+	mapstr.MergeFieldsDeep(result.Fields["process"].(mapstr.M), processMap, true)
+
+	if p.config.ReplaceFields {
+		if err := p.replaceFields(result); err != nil {
+			return nil, fmt.Errorf("replace fields: %w", err)
+		}
+	}
+	return result, nil
+}
+
+//pidToUInt32 converts PID value to uint32
+func pidToUInt32(value interface{}) (pid uint32, err error) {
+	switch v := value.(type) {
+	case string:
+		nr, err := strconv.Atoi(v)
+		if err != nil {
+			return 0, fmt.Errorf("error converting string to integer: %w", err)
+		}
+		pid = uint32(nr)
+	case uint32:
+		pid = v
+	case int, int8, int16, int32, int64:
```

Comment:
I'm curious how likely are we to get these types (and below).

Reply:
Currently, this should only be int. In the future, this processor might be used with other beats, so I tried to prepare by accepting all types for pid.

Reply:
I think here generics would be helpful to eliminate the usage of `reflect`